File: services/memory-service/app/api/routes/tools.py
"""工具端点（M-09，docs/01 §6.6）。

  GET   /api/tools                       工具清单（名称/描述/schema/风险分级）
  POST  /api/tools/execute               执行（read_only/write 直行；external 202 待确认）
  POST  /api/tools/calls/{id}/confirm    确认/拒绝 external 调用
  GET   /api/tools/calls                 调用审计（最近在前）

鉴权：Bearer JWT + workspace 成员校验（CurrentWorkspaceQuery，403 防枚举）。
M1 口径：LLM 自动 tool-loop 属 M-08 M3（LangGraph 图），本端点供前端与
调试直调；权限分级/审计/确认流已是生产口径。
"""

import uuid as uuid_mod
from typing import Annotated

from fastapi import APIRouter, Query, Response

from app.core.deps import CurrentUser, CurrentWorkspaceQuery, DbDep
from app.repositories import tool_call_repo
from app.schemas.tool import (
    ToolCallRead,
    ToolConfirmRead,
    ToolConfirmRequest,
    ToolExecuteRequest,
    ToolExecutionRead,
    ToolInfo,
)
from app.tools.builtin import doc_reader, todo, web_fetch  # noqa: F401 —— 注册副作用
from app.tools.registry import default_registry
from app.tools.router import confirm_tool_call, execute_tool

router = APIRouter(prefix="/api/tools", tags=["tools"])


@router.get("", response_model=list[ToolInfo], summary="工具清单")
async def list_tools() -> list[ToolInfo]:
    """全部已注册工具（含参数 schema 与风险分级）。"""
    return [
        ToolInfo(
            name=tool.name,
            description=tool.description,
            risk=tool.risk,
            args_schema=tool.args_model.model_json_schema() if tool.args_model else {},
        )
        for tool in default_registry.list()
    ]


@router.post(
    "/execute",
    response_model=ToolExecutionRead,
    status_code=202,
    summary="执行工具",
)
async def execute(
    ws: CurrentWorkspaceQuery,
    user: CurrentUser,
    db: DbDep,
    payload: ToolExecuteRequest,
) -> Response:
    """按风险分级执行：external 返回 202 待确认，其余 200 直接给终态。

    Raises:
        AppError: 404 工具未注册 / 422 参数不合法（RFC 7807）。
    """
    outcome = await execute_tool(
        db,
        ws_id=ws.id,
        user=user,
        name=payload.tool,
        raw_args=payload.args,
        session_id=payload.session_id,
    )
    body = ToolExecutionRead(
        call_id=str(outcome.call_id),
        status=outcome.status,
        requires_confirmation=outcome.requires_confirmation,
        result=outcome.result,
        error=outcome.error,
    )
    # external 等确认走 202；立即路径覆盖 status_code 为 200
    status_code = 202 if outcome.requires_confirmation else 200
    return Response(
        content=body.model_dump_json(exclude_none=True),
        media_type="application/json",
        status_code=status_code,
    )


@router.post("/calls/{call_id}/confirm", response_model=ToolConfirmRead, summary="确认工具调用")
async def confirm(
    call_id: uuid_mod.UUID,
    ws: CurrentWorkspaceQuery,
    user: CurrentUser,
    db: DbDep,
    payload: ToolConfirmRequest,
) -> ToolConfirmRead:
    """确认/拒绝一条 pending 的 external 调用（60s 未确认自动超时拒绝）。

    Raises:
        AppError: 404 记录不存在 / 409 已非 pending 终态。
    """
    await confirm_tool_call(db, ws_id=ws.id, user=user, call_id=call_id, approve=payload.approve)
    return ToolConfirmRead(call_id=str(call_id), accepted=True)


@router.get("/calls", response_model=list[ToolCallRead], summary="工具调用审计")
async def list_calls(
    ws: CurrentWorkspaceQuery,
    db: DbDep,
    session_id: Annotated[uuid_mod.UUID | None, Query(description="会话过滤")] = None,
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
) -> list[ToolCallRead]:
    """调用审计清单（created_at 倒序；含参数与结果摘要）。"""
    logs = await tool_call_repo.list_logs(db, ws_id=ws.id, session_id=session_id, limit=limit)
    return [ToolCallRead.model_validate(log) for log in logs]

File: frontend/src/api/tools.ts
/**
 * 工具 API（EchoDesk 前端，M-F3）。
 *
 * 对接 memory-service `/api/tools/*`：工具清单、直调执行（external 202 待确认）、
 * 确认/拒绝、调用审计。workspace_id 一律走 query 参数（CurrentWorkspaceQuery）。
 */
import { api } from "./client";

/** 工具风险分级（与服务端 ToolRiskLevel 对齐）。 */
export type ToolRisk = "read_only" | "write" | "external";

/** 调用状态（与服务端 ToolCallStatus 对齐）。 */
export type ToolCallStatus = "pending" | "success" | "denied" | "timeout" | "failed";

/** 工具清单条目（GET /api/tools）。 */
export interface ToolInfo {
  /** 工具名（如 web.fetch）。 */
  name: string;
  /** 描述。 */
  description: string;
  /** 风险分级。 */
  risk: ToolRisk;
  /** 参数 JSON Schema（MCP 同构）。 */
  args_schema: Record<string, unknown>;
}

/** 工具调用审计条目（GET /api/tools/calls）。 */
export interface ToolCall {
  /** 调用记录 ID。 */
  id: string;
  /** 来源会话（直调时可为 null）。 */
  session_id: string | null;
  /** 工具名。 */
  tool_name: string;
  /** 风险分级。 */
  risk_level: ToolRisk;
  /** 调用参数。 */
  args: Record<string, unknown>;
  /** 状态。 */
  status: ToolCallStatus;
  /** 确认人（external 调用裁决后回填）。 */
  approver_id: string | null;
  /** 结果摘要（2000 字符截断）。 */
  result_digest: string | null;
  /** 创建时间。 */
  created_at: string;
}

/** 执行工具响应体（202 pending / 200 终态）。 */
export interface ToolExecutionResult {
  /** 调用记录 ID。 */
  call_id: string;
  /** 当前状态。 */
  status: ToolCallStatus;
  /** 是否需要用户确认（external）。 */
  requires_confirmation: boolean;
  /** 执行结果（立即路径）。 */
  result: Record<string, unknown> | null;
  /** 失败原因（立即路径失败）。 */
  error: string | null;
}

/**
 * 列出全部已注册工具。
 *
 * @returns 工具清单（无 workspace 维度，全局注册表）。
 */
export async function listTools(): Promise<ToolInfo[]> {
  const { data } = await api.get<ToolInfo[]>("/api/tools");
  return data;
}

/**
 * 直调执行工具（external 风险返回 202 待确认）。
 *
 * @param workspaceId - workspace ID（query 参数）。
 * @param tool - 工具名。
 * @param args - 工具参数。
 * @param sessionId - 来源会话（审计挂链，可空）。
 * @returns 执行结果（requires_confirmation=true 时需走确认端点）。
 */
export async function executeTool(
  workspaceId: string,
  tool: string,
  args: Record<string, unknown>,
  sessionId?: string
): Promise<ToolExecutionResult> {
  const { data } = await api.post<ToolExecutionResult>(
    "/api/tools/execute",
    sessionId ? { tool, args, session_id: sessionId } : { tool, args },
    { params: { workspace_id: workspaceId } }
  );
  return data;
}

/**
 * 确认/拒绝一条 pending 的 external 调用（终态由后台任务落审计）。
 *
 * @param workspaceId - workspace ID（query 参数）。
 * @param callId - 调用记录 ID。
 * @param approve - true 执行 / false 拒绝。
 */
export async function confirmToolCall(
  workspaceId: string,
  callId: string,
  approve: boolean
): Promise<void> {
  await api.post(`/api/tools/calls/${callId}/confirm`, { approve }, {
    params: { workspace_id: workspaceId },
  });
}

/**
 * 拉取工具调用审计（created_at 倒序）。
 *
 * @param workspaceId - workspace ID（query 参数）。
 * @param sessionId - 会话过滤（可空）。
 * @param limit - 单页条数（服务端上限 100）。
 * @returns 审计清单。
 */
export async function listToolCalls(
  workspaceId: string,
  sessionId?: string,
  limit = 20
): Promise<ToolCall[]> {
  const { data } = await api.get<ToolCall[]>("/api/tools/calls", {
    params: { workspace_id: workspaceId, ...(sessionId ? { session_id: sessionId } : {}), limit },
  });
  return data;
}
